test/hotspot/jtreg/compiler/c2/irTests/TestScopedValue.java

```diff
@@ -0,0 +1,708 @@
+    private void testFastPath12Runner() throws Exception {
+        ScopedValue.where(svObject, new Object()).run(
+                () -> {
+                    Object unused = svObject.get();
+                    for (int i = 0; i < 20_000; i++) {
+                        testFastPath12();
+                        testFastPath12Inlined(0);
+                    }
+                });
+        forceCompilation("testFastPath12");
+    }
+
+    @Test
+    @IR(failOn = {IRNode.CALL_OF_METHOD, "slowGet"})
+    @IR(counts = {IRNode.IF, ">= 3", IRNode.LOAD_P_OR_N, ">= 5" })
+    @IR(counts = {IRNode.IF, "<= 4", IRNode.LOAD_P_OR_N, "<= 6" })
+    public static Object testFastPath13() {
+        // checks code shape once fully expanded
+        return svObject.get();
+    }
+
+    @Run(test = "testFastPath13", mode = RunMode.STANDALONE)
+    private void testFastPath13Runner() throws Exception {
+        ScopedValue.where(svObject, new Object()).run(
+                () -> {
+                    Object unused = svObject.get();
+                    for (int i = 0; i < 20_000; i++) {
+                        testFastPath13();
+                    }
+                });
+        forceCompilation("testFastPath13");
+    }
+
+    @Test
+    @IR(failOn = {IRNode.CALL_OF_METHOD, "slowGet"})
+    @IR(counts = {IRNode.LOAD_VECTOR_D, ">=1" })
+    public static void testFastPath14(double[] src, double[] dst) {
+        for (int i = 0; i < 10_000; i++) {
+            dst[i] = src[i] * sv.get().getValue();
+        }
+    }
+
+    @Run(test = "testFastPath14", mode = RunMode.STANDALONE)
+    private void testFastPath14Runner() throws Exception {
+        double[] array = new double[10_000];
+        ScopedValue.where(sv, new MyDouble(42)).run(
+                () -> {
+                    MyDouble unused = sv.get();
+                    for (int i = 0; i < 20_000; i++) {
+                        testFastPath14(array, array);
+                    }
+                });
+        forceCompilation("testFastPath14", double[].class, double[].class);
+    }
+
+    // Check uncommon trap is recorded at the right byte code (a cache miss) so on re-compilation,
+    // the cache null check still branches to an uncommon trap
+    @Test
+    @IR(counts = {IRNode.UNSTABLE_IF_TRAP, ">= 1" })
+    public static Object testFastPath15() {
+        return svObject.get();
+    }
+
+    @Run(test = "testFastPath15", mode = RunMode.STANDALONE)
+    private void testFastPath15Runner() throws Exception {
+        // Profile data will report a single of the 2 cache locations as a hit
+        runAndCompile15();
+        // Force a cache miss
+        ScopedValue.where(svObject, new Object()).run(
+                () -> {
+                    testFastPath15();
+                });
+        Method m = TestScopedValue.class.getDeclaredMethod("testFastPath15");
+        TestFramework.assertDeoptimizedByC2(m);
+        // Compile again
+        runAndCompile15();
```

Comment:
Might it be good to do some result verification, i.e. that the `get` always returns the expected object?